File: apps/ios_native/Sources/PasswordManageriOSCore/Sync/AutoSyncSchedulePolicy.swift
import Foundation

struct AutoSyncSchedulePolicy {
    static let syncingStatus = "Syncing..."

    var settings: SyncSettings
    var isUnlocked: Bool
    var syncStatus: String
    var lastAutoSyncAttemptAt: Date?

    var canCheckRemoteSync: Bool {
        isUnlocked &&
            settings.providerType != .none &&
            syncStatus != Self.syncingStatus
    }

    func shouldSyncOnUnlock(hasTriggeredForCurrentUnlock: Bool) -> Bool {
        !hasTriggeredForCurrentUnlock &&
            settings.autoSyncOnUnlock &&
            canCheckRemoteSync
    }

    func shouldRunIntervalSync(now: Date) -> Bool {
        guard canCheckRemoteSync,
              settings.autoSyncEnabled else {
            return false
        }

        let referenceDate = Self.maxDate(lastAutoSyncAttemptAt, settings.lastSyncAt) ?? .distantPast
        return now.timeIntervalSince(referenceDate) >= Self.intervalSeconds(for: settings)
    }

    static func intervalSeconds(for settings: SyncSettings) -> TimeInterval {
        TimeInterval(max(settings.autoSyncIntervalValue, 1)) * intervalUnitMultiplier(settings.autoSyncIntervalUnit)
    }

    private static func intervalUnitMultiplier(_ unit: SyncIntervalUnit) -> TimeInterval {
        switch unit {
        case .seconds: 1
        case .minutes: 60
        }
    }

    private static func maxDate(_ lhs: Date?, _ rhs: Date?) -> Date? {
        switch (lhs, rhs) {
        case (.none, .none): nil
        case (.some(let date), .none), (.none, .some(let date)): date
        case (.some(let lhs), .some(let rhs)): max(lhs, rhs)
        }
    }
}

File: apps/android_native/app/src/main/java/life/devops/passwordmanager/sync/AutoSyncSchedulePolicy.kt
package life.devops.passwordmanager.sync

import java.time.Instant

class AutoSyncSchedulePolicy(
    private val settings: SyncSettings,
    private val isUnlocked: Boolean,
    private val syncStatus: String,
    private val lastAutoSyncAttemptAt: Instant?,
) {
    private val canCheckRemoteSync: Boolean
        get() = isUnlocked &&
            settings.providerType != SyncProviderType.NONE &&
            syncStatus != SyncingStatus

    fun shouldSyncOnUnlock(hasTriggeredForCurrentUnlock: Boolean): Boolean =
        !hasTriggeredForCurrentUnlock &&
            settings.autoSyncOnUnlock &&
            canCheckRemoteSync

    fun shouldRunIntervalSync(now: Instant): Boolean {
        if (!canCheckRemoteSync || !settings.autoSyncEnabled) {
            return false
        }
        val reference = listOfNotNull(lastAutoSyncAttemptAt, settings.lastSyncAt).maxOrNull()
            ?: return true
        return now.toEpochMilli() - reference.toEpochMilli() >= intervalMillis(settings)
    }

    companion object {
        const val SyncingStatus = "Syncing..."

        fun intervalMillis(settings: SyncSettings): Long {
            val value = settings.autoSyncIntervalValue.coerceAtLeast(1).toLong()
            val unitMillis = when (settings.autoSyncIntervalUnit) {
                SyncIntervalUnit.SECONDS -> 1_000L
                SyncIntervalUnit.MINUTES -> 60_000L
            }
            return value * unitMillis
        }
    }
}
